File: oop/6/6/Goods.cpp
#include "Goods.h"

Goods::Goods()
	: name(NULL), deliveryPrice(0)
{
	std::cout << "Goods default constructor" << std::endl;
}

Goods::Goods(const char* name, double deliveryPrice)
	: name(new char[strlen(name) + 1]), deliveryPrice(deliveryPrice)
{
	strcpy_s(this->name, strlen(name) + 1, name);
	std::cout << "Goods constructor" << std::endl;
}

Goods::Goods(const Goods& other)
	: name(new char[strlen(other.name) + 1]), deliveryPrice(other.deliveryPrice)
{
	strcpy_s(this->name, strlen(name) + 1, name);
	std::cout << "Goods copy constructor" << std::endl;
}

Goods::~Goods() {
	if (name != NULL) delete[] name;
	std::cout << "Goods destructor" << std::endl;
}

Goods& Goods::operator=(const Goods& other) {
	if (this != &other) {
		if (name != NULL) delete[] name;
		name = new char[strlen(other.name) + 1];
		strcpy_s(name, strlen(other.name) + 1, other.name);
		deliveryPrice = other.deliveryPrice;
	}
	return *this;
}

Goods& Goods::setName(const char* name) {
	if (this->name != NULL) delete[] this->name;
	this->name = new char[strlen(name) + 1];
	strcpy_s(this->name, strlen(name) + 1, name);
	return *this;
}

Goods& Goods::setDeliveryPrice(double deliveryPrice) {
	this->deliveryPrice = deliveryPrice;
	return *this;
}

const char* Goods::getName() const {
	return name;
}

double Goods::getDeliveryPrice() const {
	return deliveryPrice;
}

File: java/course1/7/src/Goods.java
import java.io.Serializable;

public class Goods implements Serializable {
	/**
	 * 
	 */
	private static final long serialVersionUID = 4448585088855831816L;
	
	private String name;
	private double deliveryPrice;
	private double overprice;
	
	public Goods(String name, double price, double overprice) {
		this.setName(name);
		this.setDeliveryPrice(price);
		this.setOverprice(overprice);
	}

	public Goods() {
		this("", 0, 0);
	}
	
	public double getPrice() {
		return deliveryPrice + ((deliveryPrice / 100) * overprice);
	}
	
	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public double getDeliveryPrice() {
		return deliveryPrice;
	}

	public void setDeliveryPrice(double price) {
		this.deliveryPrice = price;
	}

	public double getOverprice() {
		return overprice;
	}

	public void setOverprice(double overprice) {
		this.overprice = overprice;
	}
}
